substrate-node/runtime/src/migrations/tfgrid_v15_smart_contract_v8.rs

```diff
@@ -0,0 +1,240 @@
+        debug!("migrating twin: {:?}", twin);
+
+        let new_twin = pallet_tfgrid::types::Twin::<AccountIdOf<T>> {
+            id: twin.id,
+            account_id: twin.account_id,
+            relay: None,
+            entities: twin.entities,
+            pk: None,
+        };
+        twins.insert(twin.id, new_twin);
+        reads += 1;
+    }
+
+    // Return the weight consumed by the migration.
+    return (reads, 0);
+}
+
+pub fn migrate_smart_contract<T: Config>(
+    twins: &BTreeMap<u32, pallet_tfgrid::types::Twin<AccountIdOf<T>>>,
+) -> (u64, u64) {
+    debug!(
+        " >>> Starting contract pallet migration, pallet version: {:?}",
+        pallet_smart_contract::PalletVersion::<T>::get()
+    );
+
+    let mut reads = 0;
+    let mut writes = 0;
+    let mut twin_contract_locked_balances: BTreeMap<u32, BalanceOf<T>> = BTreeMap::new();
+    // Fetch all locked balances based on the contract locks in storage and accumulate them by twin id
+    for (ctr_id, l) in pallet_smart_contract::ContractLock::<T>::iter() {
+        let ctr = pallet_smart_contract::Contracts::<T>::get(ctr_id);
+        reads += 1;
+        match ctr {
+            Some(contract) => {
+                if !twins.contains_key(&contract.twin_id) {
+                    debug!(
+                        "twins: {} does not exist, removing contract and lock",
+                        contract.twin_id
+                    );
+                    pallet_smart_contract::Contracts::<T>::remove(ctr_id);
+                    pallet_smart_contract::ContractLock::<T>::remove(ctr_id);
+                    writes += 2;
+                } else {
+                    *twin_contract_locked_balances
+                        .entry(contract.twin_id)
+                        .or_default() += l.amount_locked;
+                }
+            }
+            None => {
+                debug!(
+                    "no contract found for contract lock {}, cleaning up lock..",
+                    ctr_id
+                );
+                writes += 1;
+                pallet_smart_contract::ContractLock::<T>::remove(ctr_id);
+            }
+        }
+    }
+
+    for (twin_id, twin) in twins {
+        // If the twin needs to have some locked balance on his account because of running contracts
+        // Check how much we can actually lock based on his current total balance
+        // this will make sure the locked balance will not exceed the total balance on the twin's account
+        let should_lock = twin_contract_locked_balances.get(&twin_id).map(|b| {
+            debug!(
+                "contract locked balance on twin {} account: {:?}",
+                twin_id, b
+            );
+            (<T as Config>::Currency::total_balance(&twin.account_id)
+                - <T as Config>::Currency::minimum_balance())
+            .min(*b)
+        });
+
+        // Unlock all balance for the twin
+        <T as Config>::Currency::remove_lock(GRID_LOCK_ID, &twin.account_id);
```

Comment:
Same idea here
So all twins have a lock?

Reply:
Not at all

Reply:
I'm not sure here. We do add the lock whenever we bill something but in case nothing was ever billed I don't know. Does it even matter? Maybe @DylanVerstraete knows.

Reply:
Maybe putting this part inside the `if let Some(should_lock) = should_lock {` block ?
remove lock only when we want to set lock again

Reply:
could do that yes